File: qt/cbor/gen_qcborstreamreader.go
package cbor

import (
	"github.com/mappu/miqt/qt"
	"unsafe"
)

type QCborStreamReader__Type byte

const (
	QCborStreamReader__UnsignedInteger QCborStreamReader__Type = 0
	QCborStreamReader__NegativeInteger QCborStreamReader__Type = 32
	QCborStreamReader__ByteString      QCborStreamReader__Type = 64
	QCborStreamReader__ByteArray       QCborStreamReader__Type = 64
	QCborStreamReader__TextString      QCborStreamReader__Type = 96
	QCborStreamReader__String          QCborStreamReader__Type = 96
	QCborStreamReader__Array           QCborStreamReader__Type = 128
	QCborStreamReader__Map             QCborStreamReader__Type = 160
	QCborStreamReader__Tag             QCborStreamReader__Type = 192
	QCborStreamReader__SimpleType      QCborStreamReader__Type = 224
	QCborStreamReader__HalfFloat       QCborStreamReader__Type = 249
	QCborStreamReader__Float16         QCborStreamReader__Type = 249
	QCborStreamReader__Float           QCborStreamReader__Type = 250
	QCborStreamReader__Double          QCborStreamReader__Type = 251
	QCborStreamReader__Invalid         QCborStreamReader__Type = 255
)

type QCborStreamReader__StringResultCode int

const (
	QCborStreamReader__EndOfString QCborStreamReader__StringResultCode = 0
	QCborStreamReader__Ok          QCborStreamReader__StringResultCode = 1
	QCborStreamReader__Error       QCborStreamReader__StringResultCode = -1
)

type QCborStreamReader struct {
	h          uintptr
	isSubclass bool
}

// NewQCborStreamReader constructs a new QCborStreamReader object.
func NewQCborStreamReader() *QCborStreamReader {

	ret := newQCborStreamReader(QCborStreamReader_new())
	ret.isSubclass = true
	return ret
}

// NewQCborStreamReader2 constructs a new QCborStreamReader object.
func NewQCborStreamReader2(data string, lenVal int64) *QCborStreamReader {
	data_Cstring := CString(data)
	defer free(unsafe.Pointer(data_Cstring))

	ret := newQCborStreamReader(QCborStreamReader_new2(data_Cstring, (ptrdiff_t)(lenVal)))
	ret.isSubclass = true
	return ret
}

// NewQCborStreamReader3 constructs a new QCborStreamReader object.
func NewQCborStreamReader3(data *byte, lenVal int64) *QCborStreamReader {

	ret := newQCborStreamReader(QCborStreamReader_new3((*uchar)(unsafe.Pointer(data)), (ptrdiff_t)(lenVal)))
	ret.isSubclass = true
	return ret
}

// NewQCborStreamReader4 constructs a new QCborStreamReader object.
func NewQCborStreamReader4(data []byte) *QCborStreamReader {
	data_alias := struct_miqt_string{}
	data_alias.data = (char)(unsafe.Pointer(&data[0]))
	data_alias.len = size_t(len(data))

	ret := newQCborStreamReader(QCborStreamReader_new4(data_alias))
	ret.isSubclass = true
	return ret
}

// NewQCborStreamReader5 constructs a new QCborStreamReader object.
func NewQCborStreamReader5(device *qt.QIODevice) *QCborStreamReader {

	ret := newQCborStreamReader(QCborStreamReader_new5((*QIODevice)(device.UnsafePointer())))
	ret.isSubclass = true
	return ret
}

func (this *QCborStreamReader) SetDevice(device *qt.QIODevice) {
	QCborStreamReader_SetDevice(this.h, (*QIODevice)(device.UnsafePointer()))
}

func (this *QCborStreamReader) Device() *qt.QIODevice {
	return qt.UnsafeNewQIODevice(unsafe.Pointer(QCborStreamReader_Device(this.h)))
}

func (this *QCborStreamReader) AddData(data []byte) {
	data_alias := struct_miqt_string{}
	data_alias.data = (char)(unsafe.Pointer(&data[0]))
	data_alias.len = size_t(len(data))
	QCborStreamReader_AddData(this.h, data_alias)
}

func (this *QCborStreamReader) AddData2(data string, lenVal int64) {
	data_Cstring := CString(data)
	defer free(unsafe.Pointer(data_Cstring))
	QCborStreamReader_AddData2(this.h, data_Cstring, (ptrdiff_t)(lenVal))
}

func (this *QCborStreamReader) AddData3(data *byte, lenVal int64) {
	QCborStreamReader_AddData3(this.h, (*uchar)(unsafe.Pointer(data)), (ptrdiff_t)(lenVal))
}

func (this *QCborStreamReader) Reparse() {
	QCborStreamReader_Reparse(this.h)
}

func (this *QCborStreamReader) Clear() {
	QCborStreamReader_Clear(this.h)
}

func (this *QCborStreamReader) Reset() {
	QCborStreamReader_Reset(this.h)
}

func (this *QCborStreamReader) LastError() *QCborError {
	_goptr := newQCborError(QCborStreamReader_LastError(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QCborStreamReader) CurrentOffset() int64 {
	return (int64)(QCborStreamReader_CurrentOffset(this.h))
}

func (this *QCborStreamReader) IsValid() bool {
	return (bool)(QCborStreamReader_IsValid(this.h))
}

func (this *QCborStreamReader) ContainerDepth() int {
	return (int)(QCborStreamReader_ContainerDepth(this.h))
}

func (this *QCborStreamReader) ParentContainerType() QCborStreamReader__Type {
	return (QCborStreamReader__Type)(QCborStreamReader_ParentContainerType(this.h))
}

func (this *QCborStreamReader) HasNext() bool {
	return (bool)(QCborStreamReader_HasNext(this.h))
}

func (this *QCborStreamReader) Next() bool {
	return (bool)(QCborStreamReader_Next(this.h))
}

func (this *QCborStreamReader) Type() Type {
	xxxxxxxxx
}

func (this *QCborStreamReader) IsUnsignedInteger() bool {
	return (bool)(QCborStreamReader_IsUnsignedInteger(this.h))
}

func (this *QCborStreamReader) IsNegativeInteger() bool {
	return (bool)(QCborStreamReader_IsNegativeInteger(this.h))
}

func (this *QCborStreamReader) IsInteger() bool {
	return (bool)(QCborStreamReader_IsInteger(this.h))
}

func (this *QCborStreamReader) IsByteArray() bool {
	return (bool)(QCborStreamReader_IsByteArray(this.h))
}

func (this *QCborStreamReader) IsString() bool {
	return (bool)(QCborStreamReader_IsString(this.h))
}

func (this *QCborStreamReader) IsArray() bool {
	return (bool)(QCborStreamReader_IsArray(this.h))
}

func (this *QCborStreamReader) IsMap() bool {
	return (bool)(QCborStreamReader_IsMap(this.h))
}

func (this *QCborStreamReader) IsTag() bool {
	return (bool)(QCborStreamReader_IsTag(this.h))
}

func (this *QCborStreamReader) IsSimpleType() bool {
	return (bool)(QCborStreamReader_IsSimpleType(this.h))
}

func (this *QCborStreamReader) IsFloat16() bool {
	return (bool)(QCborStreamReader_IsFloat16(this.h))
}

func (this *QCborStreamReader) IsFloat() bool {
	return (bool)(QCborStreamReader_IsFloat(this.h))
}

func (this *QCborStreamReader) IsDouble() bool {
	return (bool)(QCborStreamReader_IsDouble(this.h))
}

func (this *QCborStreamReader) IsInvalid() bool {
	return (bool)(QCborStreamReader_IsInvalid(this.h))
}

func (this *QCborStreamReader) IsSimpleTypeWithSt(st QCborSimpleType) bool {
	return (bool)(QCborStreamReader_IsSimpleTypeWithSt(this.h, (uint8_t)(st)))
}

func (this *QCborStreamReader) IsFalse() bool {
	return (bool)(QCborStreamReader_IsFalse(this.h))
}

func (this *QCborStreamReader) IsTrue() bool {
	return (bool)(QCborStreamReader_IsTrue(this.h))
}

func (this *QCborStreamReader) IsBool() bool {
	return (bool)(QCborStreamReader_IsBool(this.h))
}

func (this *QCborStreamReader) IsNull() bool {
	return (bool)(QCborStreamReader_IsNull(this.h))
}

func (this *QCborStreamReader) IsUndefined() bool {
	return (bool)(QCborStreamReader_IsUndefined(this.h))
}

func (this *QCborStreamReader) IsLengthKnown() bool {
	return (bool)(QCborStreamReader_IsLengthKnown(this.h))
}

func (this *QCborStreamReader) Length() uint64 {
	return (uint64)(QCborStreamReader_Length(this.h))
}

func (this *QCborStreamReader) IsContainer() bool {
	return (bool)(QCborStreamReader_IsContainer(this.h))
}

func (this *QCborStreamReader) EnterContainer() bool {
	return (bool)(QCborStreamReader_EnterContainer(this.h))
}

func (this *QCborStreamReader) LeaveContainer() bool {
	return (bool)(QCborStreamReader_LeaveContainer(this.h))
}

func (this *QCborStreamReader) ReadAndAppendToString(dst string) bool {
	dst_ms := struct_miqt_string{}
	dst_ms.data = CString(dst)
	dst_ms.len = size_t(len(dst))
	defer free(unsafe.Pointer(dst_ms.data))
	return (bool)(QCborStreamReader_ReadAndAppendToString(this.h, dst_ms))
}

func (this *QCborStreamReader) ReadAndAppendToUtf8String(dst []byte) bool {
	dst_alias := struct_miqt_string{}
	dst_alias.data = (char)(unsafe.Pointer(&dst[0]))
	dst_alias.len = size_t(len(dst))
	return (bool)(QCborStreamReader_ReadAndAppendToUtf8String(this.h, dst_alias))
}

func (this *QCborStreamReader) ReadAndAppendToByteArray(dst []byte) bool {
	dst_alias := struct_miqt_string{}
	dst_alias.data = (char)(unsafe.Pointer(&dst[0]))
	dst_alias.len = size_t(len(dst))
	return (bool)(QCborStreamReader_ReadAndAppendToByteArray(this.h, dst_alias))
}

func (this *QCborStreamReader) CurrentStringChunkSize() int64 {
	return (int64)(QCborStreamReader_CurrentStringChunkSize(this.h))
}

func (this *QCborStreamReader) ToBool() bool {
	return (bool)(QCborStreamReader_ToBool(this.h))
}

func (this *QCborStreamReader) ToTag() QCborTag {
	return (QCborTag)(QCborStreamReader_ToTag(this.h))
}

func (this *QCborStreamReader) ToUnsignedInteger() uint64 {
	return (uint64)(QCborStreamReader_ToUnsignedInteger(this.h))
}

func (this *QCborStreamReader) ToNegativeInteger() QCborNegativeInteger {
	return (QCborNegativeInteger)(QCborStreamReader_ToNegativeInteger(this.h))
}

func (this *QCborStreamReader) ToSimpleType() QCborSimpleType {
	return (QCborSimpleType)(QCborStreamReader_ToSimpleType(this.h))
}

func (this *QCborStreamReader) ToFloat() float32 {
	return (float32)(QCborStreamReader_ToFloat(this.h))
}

func (this *QCborStreamReader) ToDouble() float64 {
	return (float64)(QCborStreamReader_ToDouble(this.h))
}

func (this *QCborStreamReader) ToInteger() int64 {
	return (int64)(QCborStreamReader_ToInteger(this.h))
}

func (this *QCborStreamReader) ReadAllString() string {
	var _ms struct_miqt_string = QCborStreamReader_ReadAllString(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QCborStreamReader) ReadAllUtf8String() []byte {
	var _bytearray struct_miqt_string = QCborStreamReader_ReadAllUtf8String(this.h)
	_ret := GoBytes(unsafe.Pointer(_bytearray.data), int(int64(_bytearray.len)))
	free(unsafe.Pointer(_bytearray.data))
	return _ret
}

func (this *QCborStreamReader) ReadAllByteArray() []byte {
	var _bytearray struct_miqt_string = QCborStreamReader_ReadAllByteArray(this.h)
	_ret := GoBytes(unsafe.Pointer(_bytearray.data), int(int64(_bytearray.len)))
	free(unsafe.Pointer(_bytearray.data))
	return _ret
}

func (this *QCborStreamReader) Next1(maxRecursion int) bool {
	return (bool)(QCborStreamReader_Next1(this.h, (int)(maxRecursion)))
}

File: qt/cbor/gen_qcborstreamreader.cpp
// +build ignore

#include <QByteArray>
#include <QCborError>
#include <QCborStreamReader>
#include <QIODevice>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qcborstreamreader.h>
#include "gen_qcborstreamreader.h"

#ifndef _Bool
#define _Bool bool
#endif

QCborStreamReader* QCborStreamReader_new() {
	return new QCborStreamReader();
}

QCborStreamReader* QCborStreamReader_new2(const char* data, ptrdiff_t lenVal) {
	return new QCborStreamReader(data, (qsizetype)(lenVal));
}

QCborStreamReader* QCborStreamReader_new3(const unsigned char* data, ptrdiff_t lenVal) {
	return new QCborStreamReader(static_cast<const quint8*>(data), (qsizetype)(lenVal));
}

QCborStreamReader* QCborStreamReader_new4(struct miqt_string data) {
	QByteArray data_QByteArray(data.data, data.len);
	return new QCborStreamReader(data_QByteArray);
}

QCborStreamReader* QCborStreamReader_new5(QIODevice* device) {
	return new QCborStreamReader(device);
}

void QCborStreamReader_SetDevice(QCborStreamReader* self, QIODevice* device) {
	self->setDevice(device);
}

QIODevice* QCborStreamReader_Device(const QCborStreamReader* self) {
	return self->device();
}

void QCborStreamReader_AddData(QCborStreamReader* self, struct miqt_string data) {
	QByteArray data_QByteArray(data.data, data.len);
	self->addData(data_QByteArray);
}

void QCborStreamReader_AddData2(QCborStreamReader* self, const char* data, ptrdiff_t lenVal) {
	self->addData(data, (qsizetype)(lenVal));
}

void QCborStreamReader_AddData3(QCborStreamReader* self, const unsigned char* data, ptrdiff_t lenVal) {
	self->addData(static_cast<const quint8*>(data), (qsizetype)(lenVal));
}

void QCborStreamReader_Reparse(QCborStreamReader* self) {
	self->reparse();
}

void QCborStreamReader_Clear(QCborStreamReader* self) {
	self->clear();
}

void QCborStreamReader_Reset(QCborStreamReader* self) {
	self->reset();
}

QCborError* QCborStreamReader_LastError(const QCborStreamReader* self) {
	return new QCborError(self->lastError());
}

long long QCborStreamReader_CurrentOffset(const QCborStreamReader* self) {
	qint64 _ret = self->currentOffset();
	return static_cast<long long>(_ret);
}

bool QCborStreamReader_IsValid(const QCborStreamReader* self) {
	return self->isValid();
}

int QCborStreamReader_ContainerDepth(const QCborStreamReader* self) {
	return self->containerDepth();
}

uint8_t QCborStreamReader_ParentContainerType(const QCborStreamReader* self) {
	QCborStreamReader::Type _ret = self->parentContainerType();
	return static_cast<uint8_t>(_ret);
}

bool QCborStreamReader_HasNext(const QCborStreamReader* self) {
	return self->hasNext();
}

bool QCborStreamReader_Next(QCborStreamReader* self) {
	return self->next();
}

Type QCborStreamReader_Type(const QCborStreamReader* self) {
	return self->type();
}

bool QCborStreamReader_IsUnsignedInteger(const QCborStreamReader* self) {
	return self->isUnsignedInteger();
}

bool QCborStreamReader_IsNegativeInteger(const QCborStreamReader* self) {
	return self->isNegativeInteger();
}

bool QCborStreamReader_IsInteger(const QCborStreamReader* self) {
	return self->isInteger();
}

bool QCborStreamReader_IsByteArray(const QCborStreamReader* self) {
	return self->isByteArray();
}

bool QCborStreamReader_IsString(const QCborStreamReader* self) {
	return self->isString();
}

bool QCborStreamReader_IsArray(const QCborStreamReader* self) {
	return self->isArray();
}

bool QCborStreamReader_IsMap(const QCborStreamReader* self) {
	return self->isMap();
}

bool QCborStreamReader_IsTag(const QCborStreamReader* self) {
	return self->isTag();
}

bool QCborStreamReader_IsSimpleType(const QCborStreamReader* self) {
	return self->isSimpleType();
}

bool QCborStreamReader_IsFloat16(const QCborStreamReader* self) {
	return self->isFloat16();
}

bool QCborStreamReader_IsFloat(const QCborStreamReader* self) {
	return self->isFloat();
}

bool QCborStreamReader_IsDouble(const QCborStreamReader* self) {
	return self->isDouble();
}

bool QCborStreamReader_IsInvalid(const QCborStreamReader* self) {
	return self->isInvalid();
}

bool QCborStreamReader_IsSimpleTypeWithSt(const QCborStreamReader* self, uint8_t st) {
	return self->isSimpleType(static_cast<QCborSimpleType>(st));
}

bool QCborStreamReader_IsFalse(const QCborStreamReader* self) {
	return self->isFalse();
}

bool QCborStreamReader_IsTrue(const QCborStreamReader* self) {
	return self->isTrue();
}

bool QCborStreamReader_IsBool(const QCborStreamReader* self) {
	return self->isBool();
}

bool QCborStreamReader_IsNull(const QCborStreamReader* self) {
	return self->isNull();
}

bool QCborStreamReader_IsUndefined(const QCborStreamReader* self) {
	return self->isUndefined();
}

bool QCborStreamReader_IsLengthKnown(const QCborStreamReader* self) {
	return self->isLengthKnown();
}

unsigned long long QCborStreamReader_Length(const QCborStreamReader* self) {
	quint64 _ret = self->length();
	return static_cast<unsigned long long>(_ret);
}

bool QCborStreamReader_IsContainer(const QCborStreamReader* self) {
	return self->isContainer();
}

bool QCborStreamReader_EnterContainer(QCborStreamReader* self) {
	return self->enterContainer();
}

bool QCborStreamReader_LeaveContainer(QCborStreamReader* self) {
	return self->leaveContainer();
}

bool QCborStreamReader_ReadAndAppendToString(QCborStreamReader* self, struct miqt_string dst) {
	QString dst_QString = QString::fromUtf8(dst.data, dst.len);
	return self->readAndAppendToString(dst_QString);
}

bool QCborStreamReader_ReadAndAppendToUtf8String(QCborStreamReader* self, struct miqt_string dst) {
	QByteArray dst_QByteArray(dst.data, dst.len);
	return self->readAndAppendToUtf8String(dst_QByteArray);
}

bool QCborStreamReader_ReadAndAppendToByteArray(QCborStreamReader* self, struct miqt_string dst) {
	QByteArray dst_QByteArray(dst.data, dst.len);
	return self->readAndAppendToByteArray(dst_QByteArray);
}

ptrdiff_t QCborStreamReader_CurrentStringChunkSize(const QCborStreamReader* self) {
	qsizetype _ret = self->currentStringChunkSize();
	return static_cast<ptrdiff_t>(_ret);
}

bool QCborStreamReader_ToBool(const QCborStreamReader* self) {
	return self->toBool();
}

uint64_t QCborStreamReader_ToTag(const QCborStreamReader* self) {
	QCborTag _ret = self->toTag();
	return static_cast<uint64_t>(_ret);
}

unsigned long long QCborStreamReader_ToUnsignedInteger(const QCborStreamReader* self) {
	quint64 _ret = self->toUnsignedInteger();
	return static_cast<unsigned long long>(_ret);
}

uint64_t QCborStreamReader_ToNegativeInteger(const QCborStreamReader* self) {
	QCborNegativeInteger _ret = self->toNegativeInteger();
	return static_cast<uint64_t>(_ret);
}

uint8_t QCborStreamReader_ToSimpleType(const QCborStreamReader* self) {
	QCborSimpleType _ret = self->toSimpleType();
	return static_cast<uint8_t>(_ret);
}

float QCborStreamReader_ToFloat(const QCborStreamReader* self) {
	return self->toFloat();
}

double QCborStreamReader_ToDouble(const QCborStreamReader* self) {
	return self->toDouble();
}

long long QCborStreamReader_ToInteger(const QCborStreamReader* self) {
	qint64 _ret = self->toInteger();
	return static_cast<long long>(_ret);
}

struct miqt_string QCborStreamReader_ReadAllString(QCborStreamReader* self) {
	QString _ret = self->readAllString();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QCborStreamReader_ReadAllUtf8String(QCborStreamReader* self) {
	QByteArray _qb = self->readAllUtf8String();
	struct miqt_string _ms;
	_ms.len = _qb.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _qb.data(), _ms.len);
	return _ms;
}

struct miqt_string QCborStreamReader_ReadAllByteArray(QCborStreamReader* self) {
	QByteArray _qb = self->readAllByteArray();
	struct miqt_string _ms;
	_ms.len = _qb.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _qb.data(), _ms.len);
	return _ms;
}

bool QCborStreamReader_Next1(QCborStreamReader* self, int maxRecursion) {
	return self->next(static_cast<int>(maxRecursion));
}

void QCborStreamReader_Delete(QCborStreamReader* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QCborStreamReader*>( self );
	} else {
		delete self;
	}
}
